app/frontend/wallet/shelley/shelley-transaction-planner.ts

```diff
@@ -284,59 +305,103 @@ export function computeTxPlan(
   const minimalChangeLovelace = computeMinUTxOLovelaceAmount(change.tokenBundle)
 
   // if we cannot create a change output with minimal ada we add it to the fee
-  if (change.tokenBundle.length === 0 && change.coins < minimalChangeLovelace) {
+  if (
+    allowAddingChangeToFee &&
+    change.tokenBundle.length === 0 &&
+    change.coins < minimalChangeLovelace
+  ) {
     return {
+      success: true,
+      txPlan: {
+        inputs,
+        outputs,
+        change: null,
+        certificates,
+        deposit,
+        additionalLovelaceAmount,
+        fee: (feeWithChange + change.coins) as Lovelace,
+        withdrawals,
+      },
+    }
+  }
+
+  return {
+    success: true,
+    txPlan: {
       inputs,
       outputs,
-      change: null,
+      change,
       certificates,
       deposit,
       additionalLovelaceAmount,
-      fee: (feeWithChange + change.coins) as Lovelace,
+      fee: feeWithChange,
       withdrawals,
-    }
+    },
   }
+}
 
-  // we cant build the transaction with big enough change lovelace
-  if (change.tokenBundle.length > 0 && change.coins < minimalChangeLovelace) {
-    throw new InternalError(InternalErrorReason.ChangeOutputTooSmall)
-  }
+// TODO: remove this altogether, since shelley all utxos are profitable
+export const isUtxoProfitable = (utxo: UTxO): boolean => {
```

Comment:
can't we simply remove this?

Reply:
although technically, there can still be legacy inputs that are not profitable and it would be sensible to filter them out

Reply:
It may seem from the diff that I somehow touched this code, but I haven't.. :smile: but yes lets remove it. Regarding the legacy utxos, yes there still can be legacy utxos that are not profitable but we agreed back in August last year that we are gonna use these utxos anyway since we are converting everything to staking balance. I think a user don't really want having a `Non-staking Balance` of 0.00001 ada.  Anyway Im gonna remove it if you don't have any objections.